docs/lang/articles/basic/syntax.md

```diff
@@ -4,280 +4,484 @@ sidebar_position: 1
 
-Major differences between Taichi kernels and Taichi functions are listed in the table below.
+We introduce the two terms "Taichi scope" and "Python scope" to make it easy to define the call range of a function or specify the effective range of a variable. So, before you proceed, familiarize yourself with these two terms. 
 
-| | Taichi kernels | Taichi functions |
-| :--- | :--- | :--- |
-| Can be called in | Python scope | Taichi scope |
-| Argument type annotation | Mandatory | Recommended |
-| Return type annotation | Mandatory| Recommended |
-| Return value | Scalar/Vector/Matrix | Arbitrary |
-| Max number of total elements in arguments | 8 (for OpenGL and CC) or 64 (other) | Unlimited |
-| Max number of return values in a return statement | 1 | Unlimited |
-| Max number of total elements in return values | 30 | Unlimited |
+### Taichi scope
 
+The code inside a kernel or a Taichi function is in the Taichi scope. The code in the Taichi scope is compiled by Taichi's runtime and executed in parallel on CPU or GPU devices for high-performance computation. 
 
+> Taichi scope corresponds to the *device side* in CUDA.  
 
+### Python scope
 
-## Taichi-scope vs Python-scope
+Code outside of the Taichi scope is in the Python scope. The code in the Python scope is native Python and executed by Python's virtual machine, *not* by Taichi's runtime.  
 
-Code decorated by `@ti.kernel` or `@ti.func` is in the **Taichi-scope**.
+> Python scope corresponds to the *host side* in CUDA. 
 
-They will be compiled by the Taichi compiler and executed on CPU or GPU devices in parallel with high performance.
+## Kernel
 
-:::note
-For people from CUDA, Taichi-scope = **device** side.
-:::
-
-Code outside `@ti.kernel` or `@ti.func` is in the **Python-scope**.
+A kernel is the entry point from which Taichi's runtime takes control and the smallest unit for runtime execution. You can define multiple kernels in your program, and each kernel is an *independent* child program. You call a kernel the same way you call a Python function, and you are allowed to switch back and forth between Taichi's runtime and Python's virtual machine.
 
-It is native Python code and will not be compiled by the Taichi compiler.
+Taichi's runtime compiles and executes kernels in the order you call them. It stores compiled kernels in the cache so that the next call to the same kernel does not need to be compiled again. 
 
-:::note
-For people from CUDA, Python-scope = **host** side.
+:::caution WARNING
+You must *not* call a kernel from inside another kernel or from inside a Taichi function. You can only call a kernel directly or from inside a native Python function. To put it differently, you can only call a kernel from the *Python scope*. 
 :::
 
-## Kernels
-
-A Python function decorated by `@ti.kernel` is a **Taichi kernel**:
-
-```python {1}
-@ti.kernel
-def my_kernel():
-    ...
+> A kernel corresponds to the `main()` function in C/C++, or the `__global__` function in CUDA. 
```

Comment:
nit: Remove the comparsion to `main()`.

I personally don't think Taichi kernel is analogous to  a C/C++ `main()` function: Readers may wonder whether they can only define one Taichi kernel in their program.

GPU programming is in large a different paradigm from pure CPU programming. So long as our readers can understand that GPU programming have both host and device side, that should be good enough.